### PR TITLE
[11.0][MIG] account_invoice_merge

### PR DESCRIPTION
**Pull request by ValentinVinagre**
Finished the PR https://github.com/OCA/account-invoicing/pull/365 from @ adrian-pesol

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I try to avoid the use of this module. Why do you need it?

**Comment by ValentinVinagre** [comment]
Hello @pedrobaeza,
I need this module to do a joint billing when the client needs it.
Odoo 11.0 does not provide this feature, right?

**Comment by pedrobaeza** [comment]
No, but why do you need it? If you need to merge invoices, it's because a previous flow is maybe not adjusted. (NOTE: I'm trying to avoid you problems with this module as it needs a lot of extensions depending on other installed modules).

**Comment by tuomoaura** [comment]
@pedrobaeza eCommerse solution is feeding separate sales orders to be invoiced once a month (all invoices for larger customers are merged into one). And send automatically. How would you do this without merge module?

File: account_invoice_merge/wizard/invoice_merge.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+            invs = self.env['account.invoice'].browse(ids)
+            for d in invs:
+                if d['state'] != 'draft':
+                    raise exceptions.Warning(
+                        _('At least one of the selected invoices is %s!') %
+                        d['state'])
+                if d['account_id'] != invs[0]['account_id']:
+                    raise exceptions.Warning(
+                        _('Not all invoices use the same account!'))
```
Comment (elicoidal) [inline account_invoice_merge/wizard/invoice_merge.py:37]: I would remove all "!"

File: account_invoice_merge/wizard/invoice_merge.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                if d['company_id'] != invs[0]['company_id']:
+                    raise exceptions.Warning(
+                        _('Not all invoices are at the same company!'))
```
Comment (elicoidal) [inline account_invoice_merge/wizard/invoice_merge.py]: 'Not all invoices belong to the same company'

File: account_invoice_merge/wizard/invoice_merge.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                if d['partner_id'] != invs[0]['partner_id']:
+                    raise exceptions.Warning(
+                        _('Not all invoices are for the same partner!'))
```
Comment (elicoidal) [inline account_invoice_merge/wizard/invoice_merge.py]: are from

**Comment by RoelAdriaans** [comment]
With Odoo 11.0 you can combine the sale orders for invoicing, this works without a problem.
Issues come if you have multiple sources of the invoices. 

Notable examples are the contract module and point of sale module.
Valid Use case, we have a client that sells products via the point of sale. Mostly business to consumer, but also some business to business. They don't have to pay at the till but get an invoice. Because handling a lot of invoices for just a few euro's is costly, we merge them at the end of the month and send it as a combined invoice. Other clients are invoiced at the end of the business day, or when there is a sufficient amount to invoice. 
These are all manual actions, and difficult to automate because it can change every week.. :)

File: account_invoice_merge/wizard/invoice_merge.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                if d['type'] != invs[0]['type']:
+                    raise exceptions.Warning(
+                        _('Not all invoices are of the same type!'))
+                if d['currency_id'] != invs[0]['currency_id']:
+                    raise exceptions.Warning(
+                        _('Not all invoices are at the same currency!'))
```
Comment (elicoidal) [inline account_invoice_merge/wizard/invoice_merge.py]: "The currency in the invoices is different."

File: account_invoice_merge/wizard/invoice_merge.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                if d['journal_id'] != invs[0]['journal_id']:
+                    raise exceptions.Warning(
+                        _('Not all invoices are at the same journal!'))
```
Comment (elicoidal) [inline account_invoice_merge/wizard/invoice_merge.py]: "All invoices must belong to the same journal"

File: account_invoice_merge/wizard/invoice_merge_view.xml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                        string="Are you sure you want to merge these invoices?"
+                        colspan="4" />
+                    <p>
+                        Please note that:<br/><br/>
+                        Invoices will only be merged if:<br/>
+                        * Invoices are in state 'draft'.<br/>
+                        * Invoices belong to the same partner.<br/>
+                        * Invoices have the same company, partner, address,
+                        currency, journal, salesman, account and type.<br/><br/>
+                        Lines will only be merged if:<br/>
```
Comment (elicoidal) [inline account_invoice_merge/wizard/invoice_merge_view.xml:27]: add a blank line for readibility

File: account_invoice_merge/readme/USAGE.rst [hunk 7]
```diff
@@ -0,0 +1,2 @@
+Go to a invoice tree view, select several invoices and click expand the
```
Comment (elicoidal) [inline account_invoice_merge/readme/USAGE.rst:1]: Could be improved using enumerated lists: `#. `

File: account_invoice_merge/readme/DESCRIPTION.rst [hunk 8]
```diff
@@ -0,0 +1,8 @@
+This module adds an action in the invoices list to merge invoices. Here are
+the conditions to allow merging:
+
+* Type should be the same (Customer Invoice, Supplier Invoice, Customer Refund or
+  Supplier Refund)
+* Partner should be the same
+* Currency should be the same
+* Account receivable account should be the same
```
Comment (elicoidal) [inline account_invoice_merge/readme/DESCRIPTION.rst]: reading the code there are more constraints to be added (company, currency, journal)

File: account_invoice_merge/wizard/invoice_merge.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        """
+        inv_obj = self.env['account.invoice']
+        aw_obj = self.env['ir.actions.act_window']
+        ids = self.env.context.get('active_ids', [])
+        invoices = inv_obj.browse(ids)
+        allinvoices = invoices.do_merge(keep_references=self.keep_references,
+                                        date_invoice=self.date_invoice)
+        xid = {
+            'out_invoice': 'action_invoice_tree1',
+            'out_refund': 'action_invoice_tree3',
```
Comment (elicoidal) [inline account_invoice_merge/wizard/invoice_merge.py:82]: Is it still valid in v11?

File: account_invoice_merge/wizard/invoice_merge.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+            'in_invoice': 'action_invoice_tree2',
+            'in_refund': 'action_invoice_tree4',
```
Comment (elicoidal) [inline account_invoice_merge/wizard/invoice_merge.py:84]: Is it still valid in v11?

File: account_invoice_merge/wizard/invoice_merge_view.xml [hunk 11]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
```
Comment (elicoidal) [inline account_invoice_merge/wizard/invoice_merge_view.xml:10]: add copyright

**Review by elicoidal — CHANGES_REQUESTED** [review]
Good to see that this module is surviving the years :wink:
Mostly details, no functional test yet

**Comment by pedrobaeza** [comment]
@tuomoaura you got the answer in the previous comment

@RoelAdriaans for contract module, we have develop several things for combining everything in one invoice:

* https://github.com/OCA/contract/tree/11.0/contract_sale_invoicing allows to invoice pending sales orders.
* Main module has a hook through an argument for adding to an existing invoice:

https://github.com/OCA/contract/blob/5d5387e0eda83163aec2d874bfb84cde4a651e70/contract/models/account_analytic_account.py#L264

* On v9, we developed a module for joining several contracts for same partner:

https://github.com/OCA/contract/tree/9.0/contract_invoice_merge_by_partner

In summary, it's better to unify all the invoice creation handling in one side, instead of fighting with invoice creation issues from contract and from merging invoices from other sources.

For POS, I think a better solution can be to invoice all POS sales orders together. It seems that this is more work, but doing this way, you control better which data should be put in invoices instead of delegating this in generic merge methods.

What do you think?

**Comment by MiquelRForgeFlow** [comment]
I miss the commit history of the `account_invoice_merge`.

**Comment by HaraldPanten** [comment]
I think that this module is not so useful for customer invoices, maybe. In V11, users can invoice several sale orders together as @RoelAdriaans said before. 

Even though, it would allow extra functionalities for vendor bills, right?

I don't know any kind of "purchase order combination for invoicing" action. Do you? @pedrobaeza @ValentinVinagre @tuomoaura

**Comment by ValentinVinagre** [comment]
Hi @HaraldPanten,
I see good idea develop the "purchase order combination for invoicing" action (the same thing that odoo provides with SO), if with this action and other modules we can group all type of invoicings.

**Comment by pedrobaeza** [comment]
You can add several purchase orders to one invoice using the purchase order field, or even migrate this module for mass invoicing ala sales part: https://github.com/OCA/account-invoicing/tree/9.0/purchase_batch_invoicing

**Comment by ValentinVinagre** [comment]
The module 'purchase_batch_invoicing' covers our need, but the community need the 'account_invoice_merge'?  Because with what you said with other modules and standard we can 'cover' the functionalities of this module.

**Comment by pedrobaeza** [comment]
I'm trying to push everyone for finding alternatives, but if you have already migrated the module and want to finish the PR...

**Comment by HaraldPanten** [comment]
account_invoice_merge allows the functionallity for merging common invoice lines as well. I think that user can't do it with the base order combination for invoicing action, right?

That's an important feature from my point of view.

What do you think? @ValentinVinagre @pedrobaeza

**Comment by pedrobaeza** [comment]
@HaraldPanten please read above comments where I said my PoV of this not being needed at all.

**Comment by ValentinVinagre** [comment]
@HaraldPanten It's a characteristic to keep in mind.

**Comment by cvinh** [comment]
We also use this module in the following Use case : 
- we return goods coming from several purchase order to supplier and make the corresponding credit notes according to eash initial purchase order
- the supplier sends us only one credit note including all the returning goods
So we have to merge our credit notes

File: account_invoice_merge/wizard/invoice_merge.py (hunk 1 above)
Comment (ValentinVinagre) [reply account_invoice_merge/wizard/invoice_merge.py:37]: why? Any suggestions?

**Comment by eduardgm** [comment]
@elicoidal What do you think about the last commit? Is it all correct?

**Comment by pedrobaeza** [comment]
Please preserve commit history following technical method explained in https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-11.0

**Comment by HaraldPanten** [comment]
Hi all,

Is it ready to merge?

THX!

File: account_invoice_merge/__init__.py [hunk 12]
```diff
@@ -0,0 +1,9 @@
+# Copyright 2004-2010 Tiny SPRL (http://tiny.be).
```
Comment (nikul-serpentcs) [inline account_invoice_merge/__init__.py]: @ValentinVinagre  Remove Copyright from all `__init_-.py` file.

File: account_invoice_merge/__manifest__.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+#   (http://www.eficent.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+{
+    'name': 'Account Invoice Merge',
+    'version': '11.0.1.0.0',
+    'category': 'Finance',
+    'summary': "Merge invoices in draft",
+    'author': "Elico Corp,Odoo Community Association (OCA)",
+    'website': 'http://www.openerp.net.cn',
```
Comment (nikul-serpentcs) [inline account_invoice_merge/__manifest__.py]: @ValentinVinagre URL replace as per [this](https://agithub.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L10)

File: account_invoice_merge/wizard/invoice_merge.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    _description = "Merge Partner Invoice"
+
+    keep_references = fields.Boolean('Keep references from original invoices',
+                                     default=True)
+    date_invoice = fields.Date('Invoice Date')
+
+    @api.model
+    def _dirty_check(self):
+        if self.env.context.get('active_model', '') == 'account.invoice':
+            ids = self.env.context['active_ids']
```
Comment (nikul-serpentcs) [inline account_invoice_merge/wizard/invoice_merge.py:23]: @ValentinVinagre  here check `self.env.context['active_ids', [ ]]`

**Review by nikul-serpentcs — COMMENTED** [review]
Small Change

**Comment by aidiss** [comment]
Will this ever be merged? :)

**Comment by LoisRForgeFlow** [comment]
@aidiss Seems that there are a bunch of things @ValentinVinagre has to take care of yet:
* Attend reviews.
* Attend CI checks: both travis and runbot are red.
* Preserve and clean commit history properly.

**Comment by pedrobaeza** [comment]
@aidiss you can contribute also instead of ranting.

**Comment by ValentinVinagre** [comment]
Hello, I will check the errors during this week.

**Comment by Tomaskarpovic** [comment]
@ValentinVinagre, might be time saver for you to fix this error: `ValueError: malformed node or string: <_ast.BoolOp object at 0x7f32d2a780b8>` from here: https://travis-ci.org/OCA/account-invoicing/jobs/432816208. Just don't use `or` in manifest here: https://github.com/OCA/account-invoicing/blob/ac37ee880740c48b1b2997ba3255e0b22c562d22/account_invoice_merge/__manifest__.py#L14

**Comment by aidiss** [comment]
> @aidiss you can contribute also instead of ranting.

@pedrobaeza I was trying to express interest in PR. Please do not consider my message to be a rant. Unfortunately, I do not have skills to contribute, but I do respect the work @ValentinVinagre is doing. :100:

**Comment by JordiBForgeFlow** [comment]
@tuomoaura 
> @pedrobaeza eCommerse solution is feeding separate sales orders to be invoiced once a month (all invoices for larger customers are merged into one). And send automatically. How would you do this without merge module?
The following modules would allow you to achieve the option to invoice on a monthly basis:
https://github.com/OCA/sale-workflow/tree/11.0/sale_invoice_group_method
https://github.com/OCA/sale-workflow/tree/11.0/sale_merge_draft_invoice

**Comment by bjeficent** [comment]
Hi @ValentinVinagre 
Are you planning to keep working on this?

**Comment by alvaro-domatix** [comment]
Hi, what's the status of this PR?
Even if it stills open, is there any intention to keep working on it or merge it?
@ValentinVinagre @pedrobaeza

**Comment by ValentinVinagre** [comment]
Hi @alvaro-domatix ,
I don't plan to finish one PR that has been open for almost 2 years, but you can end it if you want.
